File: src/main/java/com/chattriggers/ctjs/minecraft/libs/FileLib.java
package com.chattriggers.ctjs.minecraft.libs;

import com.chattriggers.ctjs.CTJS;
import com.chattriggers.ctjs.utils.console.Console;
import org.apache.commons.io.FileUtils;

import java.io.*;
import java.net.URL;
import java.net.URLConnection;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class FileLib {
    /**
     * Writes a file to folder in Imports.
     *
     * @param importName name of the import
     * @param fileName   name of the file
     * @param toWrite    string to write in file
     */
    public static void write(String importName, String fileName, String toWrite) {
        write(CTJS.getInstance().getConfig().getModulesFolder().value + "/" + importName + "/" + fileName, toWrite);
    }

    /**
     * Writes a file to anywhere on the system.<br>
     * Use "./" for the ".minecraft" folder.<br>
     *
     * @param fileLocation the location and file name
     * @param toWrite      string to write in file
     */
    public static void write(String fileLocation, String toWrite) {
        try {
            FileUtils.write(new File(fileLocation), toWrite);
        } catch (IOException exception) {
            Console.getConsole().printStackTrace(exception);
        }
    }

    /**
     * Reads a file from folder in Imports.<br>
     * Returns an empty string if file is not found.
     *
     * @param importName name of the import
     * @param fileName   name of the file
     * @return the string in the file
     */
    public static String read(String importName, String fileName) {
        return read(CTJS.getInstance().getConfig().getModulesFolder().value + "/" + importName + "/" + fileName);
    }

    /**
     * Reads a file from anywhere on the system.<br>
     * Use "./" for the ".minecraft" folder.<br>
     * Returns an empty string if file is not found.
     *
     * @param fileLocation the location and file name
     * @return the string in the file
     */
    public static String read(String fileLocation) {
        try {
            File file = new File(fileLocation);
            BufferedReader br = new BufferedReader(new InputStreamReader(new FileInputStream(file), "UTF8"));

            if (!file.exists() || br.readLine() == null) {
                br.close();
                return "";
            }

            br.close();
            return FileUtils.readFileToString(file);
        } catch (IOException exception) {
            Console.getConsole().printStackTrace(exception);
        }

        return "";
    }

    /**
     * Gets the contents of a url as a string
     *
     * @param theUrl the url to get the data from
     * @return the string stored in the url content
     */
    public static String getUrlContent(String theUrl) {
        StringBuilder content = new StringBuilder();

        try {
            URL url = new URL(theUrl);
            URLConnection urlConnection = url.openConnection();

            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(urlConnection.getInputStream()));

            String line;
            while ((line = bufferedReader.readLine()) != null) {
                content.append(line).append("\n");
            }
        } catch (Exception exception) {
            Console.getConsole().printStackTrace(exception);
        }

        return content.toString();
    }

    /**
     * Extracts a zip file specified by the zipFilePath to a directory specified by
     * destDirectory (will be created if does not exists)
     * @param zipFilePath the zip file path
     * @param destDirectory the destination directory
     * @throws IOException IOException
     */
    public static void unzip(String zipFilePath, String destDirectory) throws IOException {
        File destDir = new File(destDirectory);
        if (!destDir.exists()) {
            destDir.mkdir();
        }
        ZipInputStream zipIn = new ZipInputStream(new FileInputStream(zipFilePath));
        ZipEntry entry = zipIn.getNextEntry();
        // iterates over entries in the zip file
        while (entry != null) {
            String filePath = destDirectory + File.separator + entry.getName();
            if (!entry.isDirectory()) {
                // if the entry is a file, extracts it
                extractFile(zipIn, filePath);
            } else {
                // if the entry is a directory, make the directory
                File dir = new File(filePath);
                dir.mkdir();
            }
            zipIn.closeEntry();
            entry = zipIn.getNextEntry();
        }
        zipIn.close();
    }

    // helper method for unzipping
    private static void extractFile(ZipInputStream zipIn, String filePath) throws IOException {
        File toWrite = new File(filePath);
        toWrite.getParentFile().mkdirs();
        toWrite.createNewFile();

        BufferedOutputStream bos = new BufferedOutputStream(new FileOutputStream(filePath));
        byte[] bytesIn = new byte[4096];
        int read = 0;
        while ((read = zipIn.read(bytesIn)) != -1) {
            bos.write(bytesIn, 0, read);
        }
        bos.close();
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/minecraft/libs/FileLib.kt
package com.chattriggers.ctjs.minecraft.libs

import com.chattriggers.ctjs.utils.Config
import com.chattriggers.ctjs.utils.kotlin.External
import java.io.*
import java.net.URL
import java.net.UnknownHostException
import java.nio.charset.Charset
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream

@External
object FileLib {
    /**
     * Writes a file to folder in modules.
     *
     * @param importName name of the import
     * @param fileName name of the file
     * @param toWrite string to write in file
     * @param recursive whether to create folders to the file location if they don't exist
     */
    @JvmOverloads
    @JvmStatic
    fun write(importName: String, fileName: String, toWrite: String, recursive: Boolean = false) {
        write(absoluteLocation(importName, fileName), toWrite, recursive)
    }

    /**
     * Writes a file to anywhere on the system.
     * Use "./" for the ".minecraft" folder.
     *
     * @param fileLocation the location and file name
     * @param toWrite string to write in file
     * @param recursive whether to create folders to the file location if they don't exist
     */
    @JvmOverloads
    @JvmStatic
    fun write(fileLocation: String, toWrite: String, recursive: Boolean = false) {
        File(fileLocation).apply {
            if (recursive && !exists()) {
                parentFile.mkdirs()
            }
        }.writeText(toWrite)
    }

    /**
     * Writes a file to folder in modules.
     *
     * @param importName name of the import
     * @param fileName name of the file
     * @param toAppend string to append in file
     */
    @JvmStatic
    fun append(importName: String, fileName: String, toAppend: String) {
        append(absoluteLocation(importName, fileName), toAppend)
    }

    /**
     * Writes a file to anywhere on the system.
     * Use "./" for the ".minecraft" folder.
     *
     * @param fileLocation the location and file name
     * @param toAppend string to append in file
     */
    @JvmStatic
    fun append(fileLocation: String, toAppend: String) {
        File(fileLocation).appendText(toAppend)
    }

    /**
     * Reads a file from folder in modules.
     * Returns an empty string if file is not found.
     *
     * @param importName name of the import
     * @param fileName name of the file
     * @return the string in the file
     */
    @JvmStatic
    fun read(importName: String, fileName: String): String? {
        return read(File(absoluteLocation(importName, fileName)))
    }

    /**
     * Reads a file from anywhere on the system.
     * Use "./" for the ".minecraft" folder.
     * Returns an empty string if file is not found.
     *
     * @param fileLocation the location and file name
     * @return the string in the file
     */
    @JvmStatic
    fun read(fileLocation: String): String? {
        return read(File(fileLocation))
    }

    /**
     * Reads a file from anywhere on the system using java.io.File.
     *
     * @param file the java.io.File to read
     * @return the string in the file
     */
    @JvmStatic
    fun read(file: File): String? {
        return try {
            file.readText()
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Determines if a file or directory exists at the specified location
     *
     * @param importName name of the import
     * @param fileName name of the file
     * @return if the file exists
     */
    @JvmStatic
    fun exists(importName: String, fileName: String): Boolean {
        return exists(absoluteLocation(importName, fileName))
    }

    /**
     * Determines if a file or directory exists at the specified location
     *
     * @param fileLocation the path of the file
     * @return if the file exists
     */
    @JvmStatic
    fun exists(fileLocation: String): Boolean {
        return File(fileLocation).exists()
    }

    /**
     * Determines if a file or directory exists at the specified location
     *
     * @param importName name of the import
     * @param fileName name of the file
     * @return if the location is a directory
     */
    @JvmStatic
    fun isDirectory(importName: String, fileName: String): Boolean {
        return isDirectory(absoluteLocation(importName, fileName))
    }

    /**
     * Determines if a file or directory exists at the specified location
     *
     * @param fileLocation the path of the file
     * @return if the location is a directory
     */
    @JvmStatic
    fun isDirectory(fileLocation: String): Boolean {
        return File(fileLocation).isDirectory
    }

    /**
     * Gets the contents of a url as a string.
     *
     * @param theUrl the url to get the data from
     * @param userAgent the user agent to use in the connection
     * @return the string stored in the url content
     */
    @Throws(UnknownHostException::class)
    @JvmStatic
    @JvmOverloads
    fun getUrlContent(theUrl: String, userAgent: String? = "Mozilla/5.0"): String {
        val conn = URL(theUrl).openConnection()
        conn.setRequestProperty("User-Agent", userAgent)

        return conn.getInputStream().use {
            it.readBytes()
        }.toString(Charset.forName("UTF-8"))
    }

    /**
     * Deletes a file at the specified location
     *
     * @param importName name of the import
     * @param fileName name of the file
     * @return if the file was deleted
     */
    @JvmStatic
    fun delete(importName: String, fileName: String): Boolean {
        return delete(absoluteLocation(importName, fileName))
    }

    /**
     * Deletes a file at the specified location
     *
     * @param fileLocation the path of the file
     * @return if the file was deleted
     */
    @JvmStatic
    fun delete(fileLocation: String): Boolean {
        return File(fileLocation).delete()
    }

    /**
     * Deletes a directory at the specified location
     *
     * @param dir the directory to delete
     * @return if the directory was deleted
     */
    @JvmStatic
    fun deleteDirectory(dir: String): Boolean {
        return deleteDirectory(File(dir))
    }

    /**
     * Deletes a directory at the specified location
     *
     * @param dir the directory to delete
     * @return if the directory was deleted
     */
    @JvmStatic
    fun deleteDirectory(dir: File): Boolean {
        return dir.deleteRecursively()
    }

    /**
     * Extracts a zip file specified by the zipFilePath to a directory specified by
     * destDirectory (will be created if does not exist).
     * @param zipFilePath the zip file path
     * @param destDirectory the destination directory
     * @throws IOException IOException
     */
    @JvmStatic
    @Throws(IOException::class)
    fun unzip(zipFilePath: String, destDirectory: String) {
        val destDir = File(destDirectory)
        if (!destDir.exists()) destDir.mkdir()

        val zipIn = ZipInputStream(FileInputStream(zipFilePath))
        var entry: ZipEntry? = zipIn.nextEntry
        // iterates over entries in the zip file
        while (entry != null) {
            val filePath = destDirectory + File.separator + entry.name
            if (!entry.isDirectory) {
                // if the entry is a file, extracts it
                extractFile(zipIn, filePath)
            } else {
                // if the entry is a directory, make the directory
                val dir = File(filePath)
                dir.mkdir()
            }
            zipIn.closeEntry()
            entry = zipIn.nextEntry
        }
        zipIn.close()
    }

    // helper method for unzipping
    @Throws(IOException::class)
    private fun extractFile(zipIn: ZipInputStream, filePath: String) {
        val toWrite = File(filePath)
        toWrite.parentFile.mkdirs()
        toWrite.createNewFile()

        val bos = BufferedOutputStream(FileOutputStream(filePath))
        val bytesIn = ByteArray(4096)
        var read = zipIn.read(bytesIn)
        while (read != -1) {
            bos.write(bytesIn, 0, read)
            read = zipIn.read(bytesIn)
        }
        bos.close()
    }

    private fun absoluteLocation(importName: String, fileLocation: String): String {
        return Config.modulesFolder + File.separator + importName + File.separator + fileLocation
    }
}
